### PR TITLE
feat(action): introduce docker image build and push

### PR DESCRIPTION
**Pull request by IamAbbey**
This PR introduces a shared docker image build and push action for our projects.

The concluded task [[1]](https://moneymeets.atlassian.net/browse/MD-7028) had duplicated build step across relevant project.

This PR aims to solve this.

Implemented herein
- Configurable push to ECR
- Cache management
- Dynamic tagging

### CODE REVIEW

File: build-and-push-image/action.yml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  aws_region:
+    description: AWS region
+    required: false
+    default: eu-central-1
+
+  ssh_key:
+    description: SSH key
+    required: true
+  ssh_auth_sock:
+    description: SSH auth sock
```
Comment (zyv) [inline build-and-push-image/action.yml]:
```suggestion
    description: SSH auth socket
```

File: build-and-push-image/action.yml [hunk 2]
```diff
@@ -0,0 +1,100 @@
+name: Build and Push
+description: Build docker image and push to AWS ECR
+
+inputs:
+  push_to_ecr:
+    description: Depicts if the image should be pushed
```
Comment (zyv) [inline build-and-push-image/action.yml]:
```suggestion
    description: Determines whether the image should be pushed
```

**Review by zyv — COMMENTED** [review]
Just some wording suggestions.

**Review by marns93 — APPROVED** [review]
:+1: 
Nothing to add. We did a pre-review before and I think the action is good to go to use in our other repositories.

@IamAbbey 
So next step would be using this action in the repositories where you added the Dockerfiles. After we prepared the infrastructure we can update the default for `push_to_ecr` and all images will be pushed with next pipeline runs.

File: build-and-push-image/action.yml [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    description: AWS secret access key
+    required: true
+  aws_region:
+    description: AWS region
+    required: false
+    default: eu-central-1
+
+  ssh_key:
+    description: SSH key
+    required: true
```
Comment (marns93) [inline build-and-push-image/action.yml]: SSH key is optional. For instance moneymeets-backoffice is not using this.
```suggestion
    required: false
```

File: build-and-push-image/action.yml [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      uses: aws-actions/configure-aws-credentials@v4
+      with:
+        aws-access-key-id: ${{ inputs.aws_access_key_id }}
+        aws-secret-access-key: ${{ inputs.aws_secret_access_key }}
+        aws-region: ${{ inputs.aws_region }}
+
+    - id: login-ecr
+      uses: aws-actions/amazon-ecr-login@v2
+
+    - name: Setup SSH agent
```
Comment (marns93) [inline build-and-push-image/action.yml:54]: Should be executed if ssh_key is set

File: build-and-push-image/action.yml [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        file: ${{ github.workspace }}/${{ inputs.dockerfile_path }}
+        push: ${{ inputs.push_to_ecr }}
+        no-cache: ${{ inputs.push_to_ecr == 'false' }}
+        pull: true
+        tags: ${{ env.IMAGE_URL_WITH_TAG }}
+        cache-from: |
+          type=registry,ref=${{ env.IMAGE_URL }}:cache-${{ env.DOCKERFILE_HASH }}
+        cache-to: |
+          mode=max,image-manifest=true,type=registry,ref=${{ env.IMAGE_URL }}:cache-${{ env.DOCKERFILE_HASH }}
+        ssh: |
```
Comment (marns93) [inline build-and-push-image/action.yml:97]: Not needed if `ssh_key` is not set

**Review by marns93 — CHANGES_REQUESTED** [review]
I've seen that the ssh_key is required, which is not true for moneymeets-backoffice app.

**Review by marns93 — APPROVED** [review]

**Comment by marns93** [comment]
@IamAbbey We created the ECR repositories and added `APPLICATION_ID` to the repositories where you prepared the Dockerfiles.
So if you would like to, we can change the default for `push_to_ecr` to true and push the images.

In the other repository workflows you can use `ecr_repository: ${{ vars.APPLICATION_ID }}`.

**Pull request merged**